### ISSUE TITLE
Aggregations with multiple outputs

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#1796]
Some aggregations like drop_duplicates or groupby-aggregate produce dataframes or series.  Currently these output a dask.dataframe object with a single partition.  However, in cases of high cardinality this may not be optimal and we may instead want to output a multi-partition object.  

As an example, consider calling drop_duplicates on a series with very many unique objects or performing a groupby-aggregate with very many groups.  In these cases we are often bound by the tail end of the tree reduction, because the intermediate results are quite large and need to be moved many times.

We can reduce this cost by splitting the dataset in the middle of the reduction along some consistent hashing function, probably after the first reduction step.  

## Example

For drop duplicates I might do something like the following

### Bad

```python
s.drop_duplicates()  # ends with a single large partition, probably bad
```

### Good

```python
s2 = s.map_partitions(M.drop_duplicates)  # first round of reductions
h = s2.apply(myhash) % 10  # splitter
s2s = [s2[h == i] for i in range(10)]  # split to ten dask series with disjoint elements
s3s = [ss.repartition(npartitions=1) for ss in s2s]  # combine and drop_duplicates again
s4s = [ss.drop_duplicates() for ss in s3s]
s5 = dd.concat(s4s)  # bring back to single series
```

There are better ways to do this, like with `pd.cut` and such.  The above is just minimal pseudocode.

## Analysis

This produces a dask Series with 10 partitions with only unique elements.  We suffer a 10x blowup in number of tasks but spread around the data more and reduce the total depth of the reduction tree.  Communication is better spread across workers.  We did the first round of reductions first, in hopes of minimizing communication.  In principle one could imagine interleaving many rounds of splitting and re-partitioning during the tree reduction; I suspect that in practice a single round is probably sufficient up to decently large scales.

cc @jreback @jcrist

**Comment by mrocklin** [comment]
One thought would be to bake this in to `apply_concat_apply` as optional keywords that would be passed through all of the reductions.  For example we might provide two inputs:

1.  A function to determine the hashable object by which we can split.  This might be `lambda df: df.index` in the case of groupbys or `lambda df: df` in the case of `drop_duplicates`, where we need to hash the entire dataframe.
2.  Information on how many times to split the output.  This might be a list of integers in the general case, which specifies the split factor before every round of combination or single integer in a special case which splits exactly once after chunking and before the first round of combinations.

In the common case I don't think that we want this active at all.

**Comment by jcrist** [comment]
Fixed by #1808.
